src/framework/auth/strategies/password/password-strategy.ts

```diff
@@ -339,84 +292,63 @@ export class NbPasswordAuthStrategy extends NbAuthStrategy {
-              errors,
-            ));
+          return this.handleResponseError(res, module);
         }),
       );
   }
 
   refreshToken(data?: any): Observable<NbAuthResult> {
 
-    const method = this.getOption('refreshToken.method');
-    const url = this.getActionEndpoint('refreshToken');
+    const module = 'refreshToken';
+    const method = this.getOption(`${module}.method`);
+    const url = this.getActionEndpoint(module);
+    const requireValidToken = this.getOption(`${module}.requireValidToken`);
 
     return this.http.request(method, url, {body: data, observe: 'response'})
       .pipe(
         map((res) => {
-          if (this.getOption('refreshToken.alwaysFail')) {
+          if (this.getOption(`${module}.alwaysFail`)) {
             throw this.createFailResponse(data);
           }
 
           return res;
         }),
-        this.validateToken('refreshToken'),
         map((res) => {
           return new NbAuthResult(
             true,
             res,
-            this.getOption('refreshToken.redirect.success'),
+            this.getOption(`${module}.redirect.success`),
             [],
-            this.getOption('messages.getter')('refreshToken', res, this.options),
-            this.createToken(this.getOption('token.getter')('login', res, this.options)));
+            this.getOption('messages.getter')(module, res, this.options),
+            this.createToken(this.getOption('token.getter')(module, res, this.options), requireValidToken));
         }),
         catchError((res) => {
-          let errors = [];
-          if (res instanceof HttpErrorResponse) {
-            errors = this.getOption('errors.getter')('refreshToken', res, this.options);
-          } else {
-            errors.push('Something went wrong.');
-          }
-
-          return observableOf(
-            new NbAuthResult(
-              false,
-              res,
-              this.getOption('refreshToken.redirect.failure'),
-              errors,
-            ));
+          return this.handleResponseError(res, module);
         }),
       );
   }
 
-  protected validateToken(module: string): any {
-    return map((res) => {
-      const token = this.getOption('token.getter')(module, res, this.options);
-      if (!token && this.getOption(`${module}.failWhenNoToken`)) {
-        const key = this.getOption('token.key');
-        console.warn(`NbPasswordAuthStrategy:
-                          Token is not provided under '${key}' key
-                          with getter '${this.getOption('token.getter')}', check your auth configuration.`);
-
-        throw new Error('Could not extract token from the response.');
-      }
-      return res;
-    });
+  protected handleResponseError(res: any, module: string): Observable<NbAuthResult> {
+    let errors = [];
+    if (res instanceof HttpErrorResponse) {
+      errors = this.getOption('errors.getter')(module, res, this.options);
+    } else if (res instanceof NbAuthIllegalTokenError) {
```

Comment:
so in case when the token is empty any of our token classes will throw the `NbAuthTokenNotFoundError` exception, right? Which won't be caught by this `if` and will go to `something went wrong` part?

In this case the statement `It MAY be created even if backend did not return any token, in this case, it is !Valid` is not valid and `empty` token will never be created.
If my reasoning is correct, then we just need to decide, what `empty` means - malformed or invalid.

Reply:
@nnixaa here is what i have coded (or what i think i've coded 🤣 ) : 

- `NbAuthTokenNotFoundError` is thrown if we did not find the token key in the response payload (i.e. no `access_token` key in OAuth2 token response, no `token` key in simple token response...). That is what i call an `empty` token. 
=> the error is not catched during token construction, because we want to allow some endpoints to create such tokens if `requireValidToken` is false.
=> But after token construction, if `requireValidToken` is set to true, then the isValid() will return false so that the `createToken()` method will throw an `NbInvalidTokenError` saying the token is empty or not valid. `NbInvalidTokenError` will then be catched by the strategy that will display the error message.
```
if (failWhenInvalidToken && !token.isValid()) {
      // If we require a valid token (i.e. isValid), then we MUST throw NbAuthIllegalTokenError so that the strategies
      // intercept it
      throw new NbAuthIllegalTokenError('Token is empty or invalid.');
    }
```

- If we find something in the token key in the response payload (i.e `access_token` is present, `token` is present), then i consider that 
1/ we have a token (even if it is `''` or `{}`) 
2/ the token we have must respect some rules (JWT format for instance) and if not we throw NbIllegalTokenError that will be catched in every case by the strategies.


To conclude 
- `empty` means the key was not found in the response. 
- And `empty`tokens are never valid (`isValid()`) but can be created if `requireValidToken` is set to false
- If the key is found, the token is not `empty` even if its value is`''`or `{}` and then it must pass the "wellFormed" control otherwise it is declared Illegal (`NbIllegalTokenError`) and rejected.

Hope i am clear.

Reply:
I think we can decide that `''` or `{}` are empty tokens....

Reply:
Ah okay, so we catch the empty token error during this check, right?
```
    try {
      this.parsePayload();
    } catch (err) {
      if (err instanceof NbAuthIllegalTokenError) { // token is present but illegal (empty or malformed), we reject it
        throw err;
      }
    }
```
so if it is illegal - throw next, and if it's just invalid (empty) - do nothing, right?

Reply:
Exactly!
If empty the constructor does not throw any error .
Then the createToken() method (that launched the constructor) :
- do nothing if requireValidToken is false so that we have an empty token created, stored and the login/register/refresh succeeds
- throw NbIllegalTokenError (we could create NbRequireValidTokenError) so that stratégies can return the message in NBAuthResult and login/refresh/register fail with the message

Reply:
Got it, this is cool, but I'd suggest we inverse this check so that we don't block any other possible exception. So we specifically block the `NbAuthTokenNotFoundError` and don't touche other expectations including `NbAuthIllegalTokenError`:
```

 try {
      this.parsePayload();
    } catch (err) {
      if (!(err instanceof NbAuthTokenNotFoundError)) {
        throw err;
      }
    }

```

Or, another option is to not throw the `NbAuthTokenNotFoundError` expectation at all, just create the token.

Reply:
Yes I agree .
I’ m gonna push very soon

Reply:
@nnixaa What do you think now and what else can we do to be ready ?